### PR TITLE
Alonzo phase 1

### PR DESCRIPTION
**Pull request by mkv-vcm**
So like this, one commit per feature.

### CODE REVIEW

File: src/signTx.c [hunk 1]
```diff
@@ -1586,6 +1592,77 @@ static void signTx_handleMintAPDU(uint8_t p2, uint8_t* wireDataBuffer, size_t wi
 	signTxMint_handleAPDU(p2, wireDataBuffer, wireDataSize);
 }
 
+// ========================= SCRIPT DATA HASH ==========================
+
+
+enum {
+	HANDLE_DATUM_HASH_STEP_DISPLAY = 1200,
+	HANDLE_DATUM_HASH_STEP_RESPOND,
+	HANDLE_DATUM_HASH_STEP_INVALID,
```
Comment (janmazak) [inline src/signTx.c:1601]: This is not about datum hash.

File: src/signTxOutput.c [hunk 2]
```diff
@@ -624,6 +711,7 @@ bool signTxOutput_isValidInstruction(uint8_t p2)
 	case APDU_INSTRUCTION_TOP_LEVEL_DATA:
 	case APDU_INSTRUCTION_ASSET_GROUP:
 	case APDU_INSTRUCTION_TOKEN:
+	case APDU_INSTRUCTION_SCRIPT_DATA_HASH:
```
Comment (janmazak) [inline src/signTxOutput.c:714]: datum hash

File: src/txHashBuilder.c [hunk 3]
```diff
@@ -222,7 +223,11 @@ void txHashBuilder_addOutput_topLevelData(
 				BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, amount);
 			}
 		}
-		builder->state = TX_HASH_BUILDER_IN_OUTPUTS;
+		if (includeDatumHash) {
+			builder->state = TX_HASH_BUILDER_IN_OUTPUTS_DATA_HASH;
```
Comment (janmazak) [inline src/txHashBuilder.c:227]: datum hash

File: src/txHashBuilder.c [hunk 4]
```diff
@@ -1184,6 +1204,9 @@ static void txHashBuilder_assertCanLeaveMint(tx_hash_builder_t* builder)
 	default:
 		ASSERT(false);
 	}
+
+	ASSERT(builder->multiassetData.remainingAssetGroups == 0);
+	ASSERT(builder->multiassetData.remainingTokens == 0);
```
Comment (janmazak) [inline src/txHashBuilder.c:1209]: This should not be moved?

**Review by janmazak — CHANGES_REQUESTED** [review]

File: src/signTx.c (hunk 1 above)
Comment (mkv-vcm) [reply src/signTx.c:1601]: Ah, the smart rename confused the two enums and automatically renamed these as well. Nice.

File: src/signTxOutput.c (hunk 2 above)
Comment (mkv-vcm) [reply src/signTxOutput.c:714]: True.

File: src/txHashBuilder.c (hunk 3 above)
Comment (mkv-vcm) [reply src/txHashBuilder.c:227]: True.

File: src/txHashBuilder.c (hunk 4 above)
Comment (mkv-vcm) [reply src/txHashBuilder.c:1209]: That's interesting. The moving back was part of the commit chain, but should have been resolved and dissolved when git encountered the rebase to the moved back (to the multisig branch) move of the asserts.
Done.

File: src/addressUtilsShelley.c [hunk 5]
```diff
@@ -131,6 +131,33 @@ bool isStakingInfoConsistentWithAddressType(const addressParams_t* addressParams
 #undef CONSISTENT_WITH
 }
 
+staking_data_source_t determineStakingChoide(address_type_t addressType)
```
Comment (janmazak) [inline src/addressUtilsShelley.c:134]:
```suggestion
staking_data_source_t determineStakingChoice(address_type_t addressType)
```

File: src/securityPolicy.c [hunk 6]
```diff
@@ -273,16 +273,22 @@ security_policy_t policyForSignTxInput()
...
+
 	ASSERT(rawAddressSize < BUFFER_SIZE_PARANOIA);
 
 	// address type and network identification
 	ASSERT(rawAddressSize >= 1);
 	const address_type_t addressType = getAddressType(rawAddressBuffer[0]);
 	const uint8_t addressNetworkId = getNetworkId(rawAddressBuffer[0]);
 
+	if (includeDatumHash) {
+		DENY_UNLESS(determineSpendingChoice(addressType) == SPENDING_SCRIPT_HASH || determineStakingChoide(addressType) == STAKING_SCRIPT_HASH);
```
Comment (janmazak) [inline src/securityPolicy.c:289]: Perhaps would be more readable if the boolean condition got a name.
```suggestion
		const bool isScriptAddress = 
				determineSpendingChoice(addressType) == SPENDING_SCRIPT_HASH ||
				determineStakingChoide(addressType) == STAKING_SCRIPT_HASH;
		DENY_UNLESS(isScriptAddress);
```

Comment (janmazak) [reply src/securityPolicy.c:289]: Also below. Perhaps worth to make isScriptAddress a function.

File: src/addressUtilsShelley.c (hunk 5 above)
Comment (mkv-vcm) [reply src/addressUtilsShelley.c:134]: Yeah.

File: src/securityPolicy.c [hunk 7]
```diff
@@ -276,9 +253,9 @@ security_policy_t policyForSignTxInit(
 		ASSERT(false);
 	}
 
-	WARN_IF(!is_tx_network_verifiable(numOutputs, numWithdrawals, txSigningMode));
+	DENY_UNLESS(isValidNetworkId(networkId));
```
Comment (davidmisiak) [inline src/securityPolicy.c:256]: Since this is a `DENY` policy, it should be moved above the switch I think.

Comment (mkv-vcm) [reply src/securityPolicy.c:256]: Good eye, yes, we have a bunch of these lying around, especially in the calling side of the policies.

**Comment by janmazak** [comment]
Will be merged within #122

**Pull request closed without merging**